src/kernel/state/vatState.js

```diff
@@ -0,0 +1,327 @@
+    Nat(slot.id);
+    insist(
+      allowedKernelSlotTypes.includes(slot.type),
+      `unknown slot.type in '${JSON.stringify(slot)}'`,
+    );
+  }
+
+  function getKernelSlotTypedMapAndKey(kernelSlot) {
+    insistKernelSlot(kernelSlot);
+
+    const { type, id } = kernelSlot;
+    let typedKernelToVatMap;
+    let kernelSlotKey;
+
+    switch (type) {
+      case 'promise': {
+        typedKernelToVatMap = state.kernelSlotToVatSlot.promises;
+        kernelSlotKey = `${type}-${id}`;
+        break;
+      }
+      case 'resolver': {
+        typedKernelToVatMap = state.kernelSlotToVatSlot.resolvers;
+        kernelSlotKey = `${type}-${id}`;
+        break;
+      }
+      case 'export': {
+        typedKernelToVatMap = state.kernelSlotToVatSlot.exports;
+        kernelSlotKey = `${type}-${kernelSlot.vatID}-${id}`;
+        break;
+      }
+      case 'device': {
+        typedKernelToVatMap = state.kernelSlotToVatSlot.devices;
+        kernelSlotKey = `${type}-${kernelSlot.vatID}-${id}`;
+        break;
+      }
+      default:
+        throw new Error(`unexpected kernelSlot type ${kernelSlot.type}`);
+    }
+    return {
+      typedKernelToVatMap,
+      kernelSlotKey,
+    };
+  }
+
+  function getValSlotTypedMapAndKey(vatSlot) {
+    insistVatSlot(vatSlot);
+    const { type, id } = vatSlot;
+    let typedVatToKernelMap;
+    // imports, deviceImports, promises, resolvers
+
+    switch (type) {
+      case 'import': {
+        typedVatToKernelMap = state.vatSlotToKernelSlot.imports;
+        break;
+      }
+      case 'deviceImport': {
+        typedVatToKernelMap = state.vatSlotToKernelSlot.deviceImports;
+        break;
+      }
+      case 'promise': {
+        typedVatToKernelMap = state.vatSlotToKernelSlot.promises;
+        break;
+      }
+      case 'resolver': {
+        typedVatToKernelMap = state.vatSlotToKernelSlot.resolvers;
+        break;
+      }
+      default:
+        throw new Error(`unexpected vatSlot type ${vatSlot.type}`);
+    }
+    return {
+      typedVatToKernelMap,
+      vatSlotKey: `${type}-${id}`,
+    };
+  }
+
+  function getVatSlotTypeFromKernelSlot(kernelSlot) {
+    switch (kernelSlot.type) {
+      case 'export':
+        return 'import';
+      case 'device':
+        return 'deviceImport';
+      case 'promise':
+        return 'promise';
+      case 'resolver':
+        return 'resolver';
+      default:
+        throw new Error('unrecognized kernelSlot type');
+    }
+  }
+
+  function allocateNextVatSlotID(kernelSlot) {
+    const vatSlotType = getVatSlotTypeFromKernelSlot(kernelSlot);
+
+    insist(
+      allowedVatSlotTypes.includes(vatSlotType),
+      `type ${vatSlotType} is not an allowed vat slot type`,
+    );
+
+    const i = state.nextIDs[vatSlotType];
+
+    state.nextIDs[vatSlotType] = i + 1;
+
+    return i;
+  }
+
+  function mapVatSlotToKernelSlot(vatSlot) {
+    const { typedVatToKernelMap, vatSlotKey } = getValSlotTypedMapAndKey(
```

Comment:
Something about the verbose names is making this hard for me to read.. maybe it's just the wrapping to multiple lines. If we named these `map` and `key`, would that feel too terse?

And should the function be named `getVal...` or `getVat..`?

Reply:
looks good, thanks. But I'm still wondering if the `Val` in that function name is a typo of `Vat`. If it's meant to indicate a "value" of some sort, totally cool, but it seemed out of place so I wanted to ask.